File: 1730.cpp
#include <iostream>
#include <string>
#include <queue>
#include <map>
using namespace std;
#define COL first
#define ROW second
#define MAX_N 10


int n;
pair<int, int> START_POINT = {1, 1};
const char NOT_PASSED = 46;
const char ONLY_VERT = 124;
const char ONLY_HORIZON = 45;
const char BOTH_DIR = 43;
char matrix[MAX_N + 1][MAX_N + 1];
queue<char> qMovements;

void input() {
    cin >> n;
    string inp_str;
    cin >> inp_str;
    for (char e : inp_str) {
        qMovements.push(e);
    }
}
void compute() {
    for (int col = 1; col < n + 1; col++) {
        for (int row = 1; row < n + 1; row++) {
            matrix[col][row] = NOT_PASSED;
        }
    }

    map<char, map<string, int>> cmdTable {
        {'U', {{"col", -1}, {"row", 0},  {"isVertical", 1}}},
        {'D', {{"col", 1},  {"row", 0},  {"isVertical", 1}}},
        {'L', {{"col", 0},  {"row", -1}, {"isVertical", 0}}},
        {'R', {{"col", 0},  {"row", 1},  {"isVertical", 0}}}
    };
    map<char, char> mVerticalUpdate {
        make_pair(NOT_PASSED,   ONLY_VERT),
        make_pair(ONLY_VERT,    ONLY_VERT),
        make_pair(ONLY_HORIZON, BOTH_DIR),
        make_pair(BOTH_DIR,     BOTH_DIR)
    };
    map<char, char> mHorizontalUpdate {
        make_pair(NOT_PASSED,   ONLY_HORIZON),
        make_pair(ONLY_VERT,    BOTH_DIR),
        make_pair(ONLY_HORIZON, ONLY_HORIZON),
        make_pair(BOTH_DIR,     BOTH_DIR)
    };

    pair<int, int> next, current = START_POINT;
    while (!qMovements.empty()) {
        char move = qMovements.front();
        qMovements.pop();
        next = make_pair(current.COL + cmdTable[move]["col"], current.ROW + cmdTable[move]["row"]);
        if (next.COL < 1 || next.COL > n || next.ROW < 1 || next.ROW > n)
            continue;
        
        bool isVertical = cmdTable[move]["isVertical"];
        if (isVertical) {
            matrix[current.COL][current.ROW] = mVerticalUpdate[matrix[current.COL][current.ROW]];
            matrix[next.COL][next.ROW] = mVerticalUpdate[matrix[next.COL][next.ROW]];
        } else {
            matrix[current.COL][current.ROW] = mHorizontalUpdate[matrix[current.COL][current.ROW]];
            matrix[next.COL][next.ROW] = mHorizontalUpdate[matrix[next.COL][next.ROW]];
        }

        current = next;
    }
}
void output() {
    for (int col = 1; col < n + 1; col++) {
        for (int row = 1; row < n + 1; row++) {
            cout << matrix[col][row];
        }
        cout << '\n';
    }
}
int main() {
    input();
    compute();
    output();
    return 0;
}

File: 1730.py
import sys
from collections import deque


input = sys.stdin.readline
START_POINT = (1, 1)

cmdTable = {
    'U': {"col": -1, "row": 0,  "isVertical": True},
    'D': {"col": 1,  "row": 0,  "isVertical": True},
    'L': {"col": 0,  "row": -1, "isVertical": False},
    'R': {"col": 0,  "row": 1,  "isVertical": False}
}
COL = 0
ROW = 1

NOT_PASSED = 46
ONLY_VERT = 124
ONLY_HORIZON = 45
BOTH_DIR = 43


def leaveTrace(isVertical, currentTrace):
    if isVertical:
        return {chr(NOT_PASSED): chr(ONLY_VERT),
                chr(ONLY_VERT): chr(ONLY_VERT),
                chr(ONLY_HORIZON): chr(BOTH_DIR),
                chr(BOTH_DIR): chr(BOTH_DIR)}.get(currentTrace)
    else:
        return {chr(NOT_PASSED): chr(ONLY_HORIZON),
                chr(ONLY_VERT): chr(BOTH_DIR),
                chr(ONLY_HORIZON): chr(ONLY_HORIZON),
                chr(BOTH_DIR): chr(BOTH_DIR)}.get(currentTrace)


def solution(n, movements):
    matrix = [[chr(NOT_PASSED) for i in range(n+1)] for j in range(n+1)]
    # idx : 1~n
    qMovements = deque(movements)
    current = START_POINT
    while qMovements:
        move = qMovements.popleft()
        next = (current[COL] + cmdTable[move]["col"], current[ROW] + cmdTable[move]["row"])
        if not (1 <= next[COL] <= n and 1 <= next[ROW] <= n):
            continue

        isVertical = cmdTable[move]["isVertical"]
        matrix[current[COL]][current[ROW]] = leaveTrace(isVertical, matrix[current[COL]][current[ROW]])
        matrix[next[COL]][next[ROW]] = leaveTrace(isVertical, matrix[next[COL]][next[ROW]])

        current = next

    for col in range(1, n+1):
        for row in range(1, n+1):
            print(matrix[col][row], end='')
        print()


if __name__ == '__main__':
    n = int(input())
    movements = input().rstrip()
    solution(n, movements)
